wip: kvserver: change ReplicaMutex to RBMutex

rafiss opened the pull request:
### go.mod, syncutil: add xsync.RBMutex dependency

xsync.RBMutex has a few differences from the syncutil.RWMutex algorithm
which could improve performance for read-heavy access:
- Fast path for readers which avoids contention on a single counter.
- A different fairness algorithm that biases more towards readers.

###  kvserver: stop using TracedLock in Replica.maybeUnquiesce

Not sure if we'll be able to merge this, but this is to make it easier
to test out RBMutex.

###  wip: kvserver: change ReplicaMutex to RBMutex

RBMutex is biased more towards readers, so could help when there is
heavy read contention.

This usage is being changed since it appeared in a mutex profile
captured by BenchmarkParallelSysbench.

There are two notable changes that are not mechanical:
- The sync.Cond used in the replica proposal buffer was replaced with a
  channel. We can't use sync.Cond since it needs to be associated with
  a sync.Locker. and we don't have one since RBMutex does not implement
  that interface.
- The allocateIndex function of the proposal buffer sometimes needs to
  flush the buffer, which involves upgrading a shared lock to an
  exclusive lock, then downgrading it back after flushing. To support
  this, I had to make this function return the new token that can be
  used by the caller of allocateIndex to release the shared lock.

### storage: use RBMutex to protect stats

informs https://github.com/cockroachdb/cockroach/issues/142625
informs https://github.com/cockroachdb/cockroach/issues/142621
Release note: None

rafiss committed:
go.mod, syncutil: add xsync.RBMutex dependency

xsync.RBMutex has a few differences from the syncutil.RWMutex algorithm
which could improve performance for read-heavy access:
- Fast path for readers which avoids contention on a single counter.
- A different fairness algorithm that biases more towards readers.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142641)
<!-- Reviewable:end -->

rafiss committed:
kvserver: stop using TracedLock in Replica.maybeUnquiesce

Not sure if we'll be able to merge this, but this is to make it easier
to test out RBMutex.

Release note: None

rafiss committed:
kvserver: change ReplicaMutex to RBMutex

RBMutex is biased more towards readers, so could help when there is
heavy read contention.

This usage is being changed since it appeared in a mutex profile
captured by BenchmarkParallelSysbench.

There are two notable changes that are not mechanical:
- The sync.Cond used in the replica proposal buffer was replaced with a
  channel. We can't use sync.Cond since it needs to be associated with
  a sync.Locker. and we don't have one since RBMutex does not implement
  that interface.
- The allocateIndex function of the proposal buffer sometimes needs to
  flush the buffer, which involves upgrading a shared lock to an
  exclusive lock, then downgrading it back after flushing. To support
  this, I had to make this function return the new token that can be
  used by the caller of allocateIndex to release the shared lock.

Release note: None

rafiss committed:
storage: use RBMutex to protect stats

Release note: None

rafiss commented:
The results showed an improvement in some `ParallelSysbench` benchmarks using SQL, and a regression on others that use KV. From a GCE worker with:
```
benchdiff --run=BenchmarkParallelSysbench --old=5e81fe31d37 --new=fa5c172df44 ./pkg/sql/tests -b --count=20 --benchtime=5000x
```

```
name                                                    old time/op    new time/op    delta
ParallelSysbench/KV/1node_remote/oltp_begin_commit-24      632ns ±24%     578ns ±24%   -8.59%  (p=0.018 n=19+17)
ParallelSysbench/SQL/1node_local/oltp_point_select-24     20.1µs ± 9%    19.6µs ±11%   -2.39%  (p=0.046 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_only-24               462µs ± 2%     455µs ± 3%   -1.38%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_remote/oltp_read_only-24        489µs ± 1%     483µs ± 1%   -1.32%  (p=0.000 n=20+19)
ParallelSysbench/SQL/1node_local/oltp_read_only-24         353µs ± 2%     349µs ± 2%   -1.27%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_local/oltp_read_write-24        695µs ± 2%     689µs ± 2%   -0.91%  (p=0.004 n=19+20)
ParallelSysbench/SQL/3node/oltp_read_write-24             1.07ms ± 3%    1.06ms ± 2%   -0.84%  (p=0.021 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_read_only-24         273µs ± 2%     272µs ± 2%   -0.53%  (p=0.047 n=19+20)
ParallelSysbench/SQL/1node_local/oltp_write_only-24        310µs ± 3%     310µs ± 2%     ~     (p=0.708 n=20+19)
ParallelSysbench/SQL/1node_local/oltp_begin_commit-24     9.17µs ± 3%    9.07µs ± 3%     ~     (p=0.072 n=19+17)
ParallelSysbench/SQL/1node_remote/oltp_write_only-24       368µs ± 3%     367µs ± 2%     ~     (p=0.429 n=20+20)
ParallelSysbench/SQL/1node_remote/oltp_read_write-24       865µs ± 3%     860µs ± 2%     ~     (p=0.065 n=19+20)
ParallelSysbench/SQL/1node_remote/oltp_point_select-24    26.8µs ± 8%    26.8µs ± 7%     ~     (p=0.573 n=20+18)
ParallelSysbench/SQL/1node_remote/oltp_begin_commit-24    10.6µs ±20%    10.8µs ±18%     ~     (p=0.925 n=20+20)
ParallelSysbench/SQL/3node/oltp_write_only-24              526µs ± 3%     529µs ± 4%     ~     (p=0.277 n=20+20)
ParallelSysbench/SQL/3node/oltp_point_select-24           27.3µs ±11%    26.7µs ±15%     ~     (p=0.091 n=20+20)
ParallelSysbench/SQL/3node/oltp_begin_commit-24           9.44µs ±15%    9.25µs ± 5%     ~     (p=0.193 n=17+17)
ParallelSysbench/KV/1node_local/oltp_read_only-24         86.7µs ± 7%    86.8µs ± 8%     ~     (p=0.771 n=20+19)
ParallelSysbench/KV/1node_local/oltp_write_only-24         190µs ± 2%     192µs ± 4%     ~     (p=0.186 n=18+20)
ParallelSysbench/KV/1node_local/oltp_begin_commit-24       620ns ±27%     626ns ±30%     ~     (p=0.817 n=19+19)
ParallelSysbench/KV/1node_remote/oltp_point_select-24     8.25µs ± 3%    8.20µs ± 3%     ~     (p=0.410 n=18+17)
ParallelSysbench/KV/3node/oltp_read_only-24                277µs ± 2%     277µs ± 2%     ~     (p=0.718 n=20+20)
ParallelSysbench/KV/3node/oltp_point_select-24            8.46µs ± 6%    8.41µs ± 3%     ~     (p=0.483 n=20+18)
ParallelSysbench/KV/3node/oltp_begin_commit-24             580ns ±28%     598ns ±26%     ~     (p=0.659 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_read_write-24        502µs ± 2%     505µs ± 2%   +0.72%  (p=0.035 n=20+18)
ParallelSysbench/KV/1node_remote/oltp_write_only-24        235µs ± 2%     237µs ± 3%   +0.94%  (p=0.022 n=19+18)
ParallelSysbench/KV/3node/oltp_read_write-24               727µs ± 2%     734µs ± 3%   +1.04%  (p=0.022 n=19+20)
ParallelSysbench/KV/1node_local/oltp_read_write-24         325µs ± 2%     330µs ± 2%   +1.44%  (p=0.002 n=18+20)
ParallelSysbench/KV/3node/oltp_write_only-24               404µs ± 3%     410µs ± 3%   +1.63%  (p=0.001 n=20+20)
ParallelSysbench/KV/1node_local/oltp_point_select-24      3.32µs ± 5%    3.42µs ± 5%   +2.96%  (p=0.007 n=17+19)

name                                                    old alloc/op   new alloc/op   delta
ParallelSysbench/SQL/1node_local/oltp_begin_commit-24     10.5kB ± 1%    10.4kB ± 3%     ~     (p=0.089 n=16+18)
ParallelSysbench/SQL/1node_remote/oltp_begin_commit-24    10.5kB ± 2%    10.5kB ± 1%     ~     (p=0.715 n=17+17)
ParallelSysbench/SQL/3node/oltp_point_select-24           26.4kB ± 2%    26.5kB ± 1%     ~     (p=0.170 n=17+17)
ParallelSysbench/SQL/3node/oltp_begin_commit-24           10.5kB ± 2%    10.5kB ± 2%     ~     (p=0.248 n=18+17)
ParallelSysbench/KV/1node_local/oltp_begin_commit-24      1.92kB ±23%    1.91kB ±23%     ~     (p=0.867 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_begin_commit-24     1.91kB ±17%    1.82kB ±28%     ~     (p=0.085 n=20+20)
ParallelSysbench/KV/3node/oltp_point_select-24            6.81kB ± 7%    7.06kB ± 8%     ~     (p=0.052 n=20+20)
ParallelSysbench/KV/3node/oltp_begin_commit-24            1.82kB ±22%    1.88kB ±26%     ~     (p=0.529 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_only-24              1.12MB ± 0%    1.12MB ± 1%   +0.25%  (p=0.022 n=16+19)
ParallelSysbench/SQL/1node_remote/oltp_read_only-24       1.11MB ± 0%    1.11MB ± 0%   +0.39%  (p=0.000 n=19+20)
ParallelSysbench/SQL/1node_local/oltp_read_only-24         806kB ± 0%     811kB ± 0%   +0.60%  (p=0.000 n=19+17)
ParallelSysbench/KV/1node_remote/oltp_read_only-24         646kB ± 0%     651kB ± 0%   +0.68%  (p=0.000 n=18+19)
ParallelSysbench/SQL/1node_local/oltp_read_write-24       1.23MB ± 0%    1.24MB ± 1%   +0.69%  (p=0.000 n=17+20)
ParallelSysbench/KV/3node/oltp_read_only-24                653kB ± 0%     658kB ± 0%   +0.71%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_remote/oltp_read_write-24      1.59MB ± 1%    1.61MB ± 1%   +0.88%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_remote/oltp_point_select-24    26.3kB ± 1%    26.5kB ± 1%   +1.08%  (p=0.000 n=17+20)
ParallelSysbench/SQL/3node/oltp_read_write-24             2.04MB ± 1%    2.06MB ± 1%   +1.08%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_local/oltp_point_select-24     23.2kB ± 0%    23.5kB ± 1%   +1.20%  (p=0.000 n=17+18)
ParallelSysbench/KV/1node_remote/oltp_read_write-24        958kB ± 0%     972kB ± 0%   +1.43%  (p=0.000 n=20+20)
ParallelSysbench/KV/3node/oltp_read_write-24              1.33MB ± 0%    1.36MB ± 0%   +1.58%  (p=0.000 n=20+18)
ParallelSysbench/KV/1node_local/oltp_read_only-24          256kB ± 1%     260kB ± 1%   +1.63%  (p=0.000 n=19+20)
ParallelSysbench/SQL/3node/oltp_write_only-24              889kB ± 2%     904kB ± 2%   +1.66%  (p=0.002 n=20+20)
ParallelSysbench/SQL/1node_remote/oltp_write_only-24       482kB ± 0%     491kB ± 0%   +1.90%  (p=0.000 n=17+17)
ParallelSysbench/KV/3node/oltp_write_only-24               597kB ± 0%     611kB ± 0%   +2.42%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_local/oltp_read_write-24         489kB ± 1%     503kB ± 1%   +2.81%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_local/oltp_write_only-24        414kB ± 2%     426kB ± 0%   +2.97%  (p=0.000 n=20+17)
ParallelSysbench/KV/1node_remote/oltp_write_only-24        288kB ± 0%     297kB ± 0%   +3.17%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_local/oltp_write_only-24         228kB ± 0%     237kB ± 0%   +3.93%  (p=0.000 n=20+19)
ParallelSysbench/KV/1node_remote/oltp_point_select-24     6.46kB ± 2%    6.78kB ± 2%   +4.91%  (p=0.000 n=19+19)
ParallelSysbench/KV/1node_local/oltp_point_select-24      4.33kB ± 2%    4.61kB ± 2%   +6.50%  (p=0.000 n=18+19)

name                                                    old allocs/op  new allocs/op  delta
ParallelSysbench/SQL/1node_remote/oltp_begin_commit-24      66.3 ± 1%      66.0 ± 0%   -0.47%  (p=0.035 n=16+15)
ParallelSysbench/SQL/1node_local/oltp_begin_commit-24       66.0 ± 0%      65.9 ± 1%     ~     (p=0.376 n=12+15)
ParallelSysbench/SQL/3node/oltp_begin_commit-24             66.5 ± 2%      66.2 ± 2%     ~     (p=0.263 n=19+18)
ParallelSysbench/KV/1node_local/oltp_begin_commit-24        3.75 ±33%      3.60 ±17%     ~     (p=0.547 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_begin_commit-24       3.80 ±32%      3.40 ±41%     ~     (p=0.077 n=20+20)
ParallelSysbench/KV/3node/oltp_begin_commit-24              3.60 ±17%      3.50 ±14%     ~     (p=0.751 n=20+20)
ParallelSysbench/SQL/3node/oltp_point_select-24              224 ± 6%       224 ± 1%   +0.14%  (p=0.024 n=20+17)
ParallelSysbench/SQL/3node/oltp_read_only-24               4.18k ± 0%     4.24k ± 1%   +1.36%  (p=0.000 n=16+20)
ParallelSysbench/SQL/1node_remote/oltp_read_only-24        4.12k ± 0%     4.19k ± 0%   +1.68%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_local/oltp_read_only-24         3.59k ± 0%     3.66k ± 0%   +2.00%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_remote/oltp_point_select-24       219 ± 0%       224 ± 0%   +2.42%  (p=0.000 n=15+17)
ParallelSysbench/SQL/1node_local/oltp_point_select-24        184 ± 0%       189 ± 1%   +2.57%  (p=0.000 n=12+19)
ParallelSysbench/SQL/1node_remote/oltp_read_write-24       7.42k ± 0%     7.65k ± 0%   +3.00%  (p=0.000 n=18+19)
ParallelSysbench/SQL/3node/oltp_read_write-24              8.64k ± 1%     8.95k ± 1%   +3.62%  (p=0.000 n=19+20)
ParallelSysbench/SQL/1node_local/oltp_read_write-24        6.08k ± 0%     6.31k ± 0%   +3.74%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_read_only-24         1.74k ± 0%     1.81k ± 0%   +4.00%  (p=0.000 n=16+15)
ParallelSysbench/KV/3node/oltp_read_only-24                1.75k ± 0%     1.82k ± 0%   +4.05%  (p=0.000 n=18+14)
ParallelSysbench/SQL/1node_remote/oltp_write_only-24       3.21k ± 0%     3.35k ± 0%   +4.48%  (p=0.000 n=19+20)
ParallelSysbench/SQL/3node/oltp_write_only-24              4.08k ± 2%     4.30k ± 2%   +5.36%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_read_write-24        3.87k ± 0%     4.09k ± 0%   +5.59%  (p=0.000 n=20+20)
ParallelSysbench/SQL/1node_local/oltp_write_only-24        2.46k ± 0%     2.60k ± 0%   +5.85%  (p=0.000 n=20+20)
ParallelSysbench/KV/3node/oltp_read_write-24               5.15k ± 0%     5.47k ± 0%   +6.17%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_write_only-24        1.99k ± 0%     2.13k ± 0%   +7.15%  (p=0.000 n=20+20)
ParallelSysbench/KV/3node/oltp_write_only-24               2.85k ± 0%     3.09k ± 1%   +8.29%  (p=0.000 n=17+20)
ParallelSysbench/KV/3node/oltp_point_select-24              51.6 ± 3%      56.5 ± 3%   +9.60%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_remote/oltp_point_select-24       51.0 ± 0%      56.2 ± 1%  +10.29%  (p=0.000 n=15+20)
ParallelSysbench/KV/1node_local/oltp_write_only-24         1.23k ± 0%     1.37k ± 1%  +11.50%  (p=0.000 n=20+20)
ParallelSysbench/KV/1node_local/oltp_read_write-24         1.76k ± 0%     1.97k ± 1%  +12.06%  (p=0.000 n=18+19)
ParallelSysbench/KV/1node_local/oltp_read_only-24            431 ± 0%       500 ± 0%  +16.03%  (p=0.000 n=18+18)
ParallelSysbench/KV/1node_local/oltp_point_select-24        21.0 ± 0%      25.4 ± 2%  +21.19%  (p=0.000 n=18+20)
```

rafiss committed:
add bench for RBMutex

rafiss committed:
change to RWMUtex for easy diff
